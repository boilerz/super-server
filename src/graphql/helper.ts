import { GraphQLError, GraphQLFormattedError } from 'graphql';

import logger from '@boilerz/logger';

export interface FormattedError extends GraphQLFormattedError {
  code?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  context?: any;
}

export function formatError(error: GraphQLError): FormattedError {
  const { originalError, message, locations, path } = error;
  const formattedError: FormattedError = { message, locations, path };

  logger.error({ formattedError, err: originalError }, '[graphql.formatError]');

  return formattedError;
}
