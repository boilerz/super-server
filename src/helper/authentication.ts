import _ from 'lodash';
import { ResolverData } from 'type-graphql';

import logger from '@boilerz/logger';

export interface DecodedToken {
  id: string;
  exp: number;
  iat: number;
  firstName: string;
  lastName: number;
  roles: string[];
}

export interface AuthCoreContext {
  accessToken: string | null;
  decodedToken: DecodedToken | null;
}

function isAccessTokenValid(
  decodedToken: DecodedToken,
  allowedRoles: string[],
): boolean {
  try {
    const { exp, roles: userRoles }: DecodedToken = decodedToken;

    if (
      !_.isEmpty(allowedRoles) &&
      !userRoles.some((userRole) => allowedRoles.includes(userRole))
    ) {
      return false;
    }

    return exp * 1000 - Date.now() > 0;
  } catch (err) {
    logger.error({ err }, '[helper.getAccessToken] Unknown');
    return false;
  }
}

export function authChecker(
  { context }: ResolverData<AuthCoreContext>,
  roles: string[],
): boolean {
  const { decodedToken } = context;

  return !!decodedToken && isAccessTokenValid(decodedToken!, roles);
}
