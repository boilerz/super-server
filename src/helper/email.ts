import url from 'url';

import mail from '@sendgrid/mail';

import PublisherClient from '@boilerz/amqp-helper/PublisherClient';

import config from '../config';
import User from '../model/user/User';

export const EXCHANGE_NAME = 'super-server-auth-core';

interface CommonMailMessage extends Pick<User, 'firstName' | 'lastName'> {
  email: string;
}

export interface EmailValidationMessage extends CommonMailMessage {
  validationCode: string;
}

export interface LinkAccountMessage extends CommonMailMessage {
  linkCode: string;
}

export type EmailMessage = EmailValidationMessage | LinkAccountMessage;

export type EmailRootingKey = 'emailValidation' | 'linkAccount';

let publisherClient: PublisherClient<EmailMessage, EmailRootingKey>;

export function getPublisherClient(): PublisherClient<
  EmailMessage,
  EmailRootingKey
> {
  return publisherClient;
}

export async function setupEmailValidationPublisherClient(): Promise<void> {
  publisherClient = await PublisherClient.createAndSetupClient<
    EmailMessage,
    EmailRootingKey
  >({
    amqpUrl: config.amqpUrl,
    exchangeName: EXCHANGE_NAME,
  });
}

export async function sendLinkAccountRequest(
  message: LinkAccountMessage,
): Promise<void> {
  await publisherClient.publish(message, 'linkAccount');
}

export async function sendLinkAccountEmail(
  message: LinkAccountMessage,
): Promise<void> {
  await mail.send({
    hideWarnings: true,
    templateId: config.sendgrid.mailTemplates.linkAccountId,
    from: config.sendgrid.senderEmail,
    to: message.email,
    dynamicTemplateData: {
      firstName: message.firstName,
      lastName: message.lastName,
      linkCode: message.linkCode,
    },
  });
}

export async function sendValidationEmailRequest(
  message: EmailValidationMessage,
): Promise<void> {
  await publisherClient.publish(message, 'emailValidation');
}

export async function sendValidationEmail(
  message: EmailValidationMessage,
): Promise<void> {
  await mail.send({
    hideWarnings: true,
    templateId: config.sendgrid.mailTemplates.emailValidationId,
    from: config.sendgrid.senderEmail,
    to: message.email,
    dynamicTemplateData: {
      firstName: message.firstName,
      lastName: message.lastName,
      validationUrl: url.format({
        host: config.host,
        query: {
          email: message.email,
          validationCode: message.validationCode,
        },
      }),
    },
  });
}
