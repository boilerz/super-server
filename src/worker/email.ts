import logger from '@boilerz/logger';
import ConsumerClient from '@boilerz/amqp-helper/ConsumerClient';
import mail from '@sendgrid/mail';
import assert from 'assert';
import {
  EXCHANGE_NAME,
  EmailValidationMessage,
  sendValidationEmail,
  EmailMessage,
  EmailRootingKey,
  LinkAccountMessage,
  sendLinkAccountEmail,
} from '../helper/email';
import config from '../config';

let consumerClient: ConsumerClient<EmailValidationMessage>;

export function getConsumerClient(): ConsumerClient<EmailValidationMessage> {
  return consumerClient;
}

export async function start(): Promise<void> {
  assert(config.host, 'SERVER_HOST need to be set');
  assert(config.sendgrid.apiKey, 'SENDGRID_API_KEY need to be set');
  assert(config.sendgrid.senderEmail, 'SENDER_EMAIL need to be set');
  assert(
    config.sendgrid.mailTemplates.emailValidationId,
    'EMAIL_VALIDATION_TEMPLATE_ID need to be set',
  );

  mail.setApiKey(config.sendgrid.apiKey);

  consumerClient = await ConsumerClient.createAndSetupClient<
    EmailMessage,
    EmailRootingKey
  >({
    amqpUrl: config.amqpUrl,
    exchangeName: EXCHANGE_NAME,
    queueName: 'email',
    nAckThrottle: 5000,
    onMessageHandlerByRootingKey: {
      async emailValidation(message: EmailValidationMessage): Promise<void> {
        logger.info({ message }, '[worker.email] Email validation');
        await sendValidationEmail(message);
      },
      async linkAccount(message: LinkAccountMessage): Promise<void> {
        logger.info({ message }, '[worker.email] Account linking');
        await sendLinkAccountEmail(message);
      },
    },
  });

  await consumerClient.consume();
}

export async function shutdown(signalOrError: string | Error): Promise<void> {
  if (signalOrError instanceof Error) {
    logger.error({ err: signalOrError }, '[shutdown] Uncaught exception');
  } else {
    logger.info(
      { signal: signalOrError },
      '[shutdown] Shutdown signal received',
    );
  }

  try {
    await consumerClient?.tearDown();
  } catch (err) {
    logger.error({ err }, '[shutdown] Shutdown failure');
  } finally {
    process.exit(0);
  }
}

/* istanbul ignore if */
if (!module.parent) {
  process
    .on('SIGTERM', shutdown)
    .on('SIGINT', shutdown)
    .on('uncaughtException', shutdown);

  start().then(() => logger.info('[emailValidation] - Started'));
}
