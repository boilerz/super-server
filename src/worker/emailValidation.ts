import logger from '@boilerz/logger';
import ConsumerClient from '@boilerz/amqp-helper/ConsumerClient';
import {
  EXCHANGE_NAME,
  EmailValidationMessage,
  sendValidationEmail,
} from '../helper/email';

let consumerClient: ConsumerClient<EmailValidationMessage>;

export function getConsumerClient(): ConsumerClient<EmailValidationMessage> {
  return consumerClient;
}

export async function start(): Promise<void> {
  consumerClient = await ConsumerClient.createAndSetupClient<
    EmailValidationMessage
  >({
    exchangeName: EXCHANGE_NAME,
    queueName: 'email-validation',
    nAckThrottle: 5000,
    async onMessageHandler(message, rootingKey): Promise<void> {
      logger.info({ message, rootingKey }, 'single handler consumer');
      await sendValidationEmail(message);
    },
  });

  await consumerClient.consume();
}

export async function shutdown(signalOrError: string | Error): Promise<void> {
  if (signalOrError instanceof Error) {
    logger.error({ err: signalOrError }, '[shutdown] Uncaught exception');
  } else {
    logger.info(
      { signal: signalOrError },
      '[shutdown] Shutdown signal received',
    );
  }

  try {
    await consumerClient?.tearDown();
  } catch (err) {
    logger.error({ err }, '[shutdown] Shutdown failure');
  } finally {
    process.exit(0);
  }
}

/* istanbul ignore if */
if (!module.parent) {
  process
    .on('SIGTERM', shutdown)
    .on('SIGINT', shutdown)
    .on('uncaughtException', shutdown);

  start().then(() => logger.info('[emailValidation] - Started'));
}
