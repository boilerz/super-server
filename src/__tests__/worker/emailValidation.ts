import mail from '@sendgrid/mail';
import * as emailWorker from '../../worker/email';
import * as emailHelper from '../../helper/email';

describe('[worker] emailValidation', () => {
  let processExitSpy: jest.SpyInstance;
  const message = {
    email: 'john@doe.com',
    firstName: 'John',
    lastName: 'Doe',
  };
  beforeAll(async () => {
    await emailHelper.setupEmailValidationPublisherClient();
  });
  afterAll(async () => {
    await emailHelper.getPublisherClient().tearDown();
  });

  beforeEach(() => {
    processExitSpy = jest.spyOn(process, 'exit').mockReturnThis();
  });

  it('should handle a send validation email request', async () => {
    const mailSendSpy = jest.spyOn(mail, 'send').mockReturnThis();

    await emailWorker.start();
    await emailHelper.sendValidationEmailRequest({
      ...message,
      validationCode: '42',
    });
    await emailWorker.getConsumerClient().waitEmptiness();
    await emailWorker.shutdown('SIGINT');

    expect(processExitSpy).toHaveBeenCalledWith(0);
    expect(mailSendSpy).toHaveBeenCalledWith({
      dynamicTemplateData: {
        firstName: 'John',
        lastName: 'Doe',
        validationUrl:
          'http://localhost:3000?email=john%40doe.com&validationCode=42',
      },
      from: 'john.sender@doe.co',
      hideWarnings: true,
      templateId: 'dummy-email-validation-id',
      to: 'john@doe.com',
    });
  });

  it('should handle a send link code email request', async () => {
    const mailSendSpy = jest.spyOn(mail, 'send').mockReturnThis();

    await emailWorker.start();
    await emailHelper.sendLinkAccountRequest({
      ...message,
      linkCode: '15',
    });
    await emailWorker.getConsumerClient().waitEmptiness();
    await emailWorker.shutdown('SIGINT');

    expect(processExitSpy).toHaveBeenCalledWith(0);
    expect(mailSendSpy).toHaveBeenCalledWith({
      dynamicTemplateData: {
        firstName: 'John',
        lastName: 'Doe',
        linkCode: '15',
      },
      from: 'john.sender@doe.co',
      hideWarnings: true,
      templateId: 'dummy-link-account-id',
      to: 'john@doe.com',
    });
  });
});
