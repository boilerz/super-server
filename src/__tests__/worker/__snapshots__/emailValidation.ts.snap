// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[worker] emailValidation should handle a send link code email request 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "dynamicTemplateData": Object {
          "firstName": "John",
          "lastName": "Doe",
          "linkCode": "15",
        },
        "from": "john.sender@doe.co",
        "hideWarnings": true,
        "templateId": "dummy-link-account-id",
        "to": "john@doe.com",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": MailService {
        "MailService": [Function],
        "client": Client {
          "auth": "Bearer SG.dummy-api-key",
          "defaultHeaders": Object {
            "Accept": "application/json",
            "Content-Type": "application/json",
            "User-Agent": "sendgrid/7.2.0;nodejs",
          },
          "defaultRequest": Object {
            "baseUrl": "https://api.sendgrid.com/",
            "headers": Object {},
            "maxContentLength": Infinity,
            "method": "GET",
            "url": "",
          },
          "impersonateSubuser": "",
        },
        "secretRules": Array [],
        "send": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "dynamicTemplateData": Object {
                  "firstName": "John",
                  "lastName": "Doe",
                  "linkCode": "15",
                },
                "from": "john.sender@doe.co",
                "hideWarnings": true,
                "templateId": "dummy-link-account-id",
                "to": "john@doe.com",
              },
            ],
          ],
          "results": [Circular],
        },
        "substitutionWrappers": Array [
          "{{",
          "}}",
        ],
      },
    },
  ],
}
`;

exports[`[worker] emailValidation should handle a send validation email request 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "dynamicTemplateData": Object {
          "firstName": "John",
          "lastName": "Doe",
          "validationUrl": "http://localhost:3000?email=john%40doe.com&validationCode=42",
        },
        "from": "john.sender@doe.co",
        "hideWarnings": true,
        "templateId": "dummy-email-validation-id",
        "to": "john@doe.com",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": MailService {
        "MailService": [Function],
        "client": Client {
          "auth": "Bearer SG.dummy-api-key",
          "defaultHeaders": Object {
            "Accept": "application/json",
            "Content-Type": "application/json",
            "User-Agent": "sendgrid/7.4.1;nodejs",
          },
          "defaultRequest": Object {
            "baseUrl": "https://api.sendgrid.com/",
            "headers": Object {},
            "maxBodyLength": Infinity,
            "maxContentLength": Infinity,
            "method": "GET",
            "url": "",
          },
          "impersonateSubuser": "",
        },
        "secretRules": Array [],
        "send": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "dynamicTemplateData": Object {
                  "firstName": "John",
                  "lastName": "Doe",
                  "validationUrl": "http://localhost:3000?email=john%40doe.com&validationCode=42",
                },
                "from": "john.sender@doe.co",
                "hideWarnings": true,
                "templateId": "dummy-email-validation-id",
                "to": "john@doe.com",
              },
            ],
          ],
          "results": [Circular],
        },
        "substitutionWrappers": Array [
          "{{",
          "}}",
        ],
      },
    },
  ],
}
`;
